File: Parallelized_SEIR_model.jl
using Distributions
using Random
using LinearAlgebra
using StatsBase
using Base.Threads
using DelimitedFiles
using PyCall

function p_from_rate(x::Array,delta_t)
    # Compute the probability of transition from the rate in the case of the multi-patch model
    
    return -exp.(-x .* delta_t) .+ 1
end

function p_from_rate_single_patch(x::Float64,delta_t)
    # Compute the probability of transition from the rate in the case of the single-patch model
    
    return -exp(-x * delta_t) + 1
end

function force_infection(map::Array, I::Array, beta::Float64)
    # Compute the force of infection for every tile of the multi-patch model
    
    return map * I * beta
end

 function SEIRS_multi_patch!(E::Array,I::Array,R::Array,T,delta_t,N::Array,parameters, map::Array)
#= T is the total observation time (days)
   delta_t is the time-step (days)
   N is the array of total population in study in each tile
   parameters is a dictionary of values for the parameters defining the mathematical process 
   map is the colocation map used =#

    p_mu=p_from_rate_single_patch(parameters["mu"],delta_t)
    p_alpha=p_from_rate_single_patch(parameters["alpha"],delta_t)
    p_omega=p_from_rate_single_patch(parameters["omega"],delta_t)
    
    for t = range(1,convert(Int32,round(T/delta_t))-1)
        p_beta=p_from_rate(force_infection(map[1+convert(Int32,floor((t-1)/convert(Int32,round(7/dt)))),:,:].*transpose(N),I[t,:]./N,parameters["beta"]),delta_t)
	    for i = range(1,size(E,2))
            E_new=rand(Binomial(N[i] - E[t,i] - I[t,i] - R[t,i],p_beta[i]))
            I_new=rand(Binomial(E[t,i],p_alpha))
            R_new=rand(Binomial(I[t,i],p_mu))
            S_new=rand(Binomial(R[t,i],p_omega))
            E[t+1,i]=E[t,i]+E_new-I_new
            I[t+1,i]=I[t,i]+I_new-R_new
            R[t+1,i]=R[t,i]+R_new-S_new
        end
        #if ((sum(I[t+1,:])+sum(E[t+1,:]))==0)
         #   break
        #end
    end
end

#function SEIRS_ODE_multi!(du,u,params,t)
    #= ODE version of the multi-patch stochastic model:
       p[1] is the beta rate (S-->E),
       p[2] is the alpha rate (E-->I),
       p[3] is the mu rate (I-->R),
       p[4] is the omega rate (R-->S),
       N is the total population,
       map is the colocation map between tiles =#
    
    #N,map,p=params
    #force_of_infection=force_infection(map.*N,u[2,:]./N,p[1])    
    #Tiles=size(map,1)
    #for i = range(1,Tiles) 
    #    S=N[i]-u[1,i]-u[2,i]-u[3,i]
    #    du[1,i] = force_of_infection[i]*S - p[2]*u[1,i]
    #    du[2,i] = p[2]*u[1,i] - p[3]*u[2,i]
    #    du[3,i] = p[3]*u[2,i] - p[4]*u[3,i]
    #end
#end 

starting=time()

Iterations_multi=parse(Int32,ARGS[1])

include(ARGS[2])
address=address_colocation
dir=readdir(address)
coloc_3d=zeros(Float64,size(dir,1)-1,110,110)
coloc_baseline=zeros(Float64,110,110)
for i in 1:size(dir,1)
    if (endswith(dir[i],"_baseline.txt"))
        coloc=readdlm("$address/$(dir[i])", ',', '\n')
        coloc=replace(coloc, NaN => 0)
        coloc_baseline.=coloc
    elseif (endswith(dir[i],".txt"))
        coloc = readdlm("$address/$(dir[i])", ',', '\n')
        coloc=replace(coloc, NaN => 0)
        coloc_3d[i,:,:].=coloc
    end
end
N=Int32.(readdlm(address_population, ',', '\n'))
 
open_file = open(dictionary_address, "r")
dict_str = read(open_file, String)
close(open_file)

# Parse the dictionary string using Python's ast module
python = pyimport("ast")
dict_data = python.literal_eval(dict_str)


println("Reading part successful")

params = include("params.jl")
index_seeding=dict_data[params["Seeding_department"]]
time_seeding=1+(Date(params["time_seeding"])-Date("2023-03-27")).value
params["R_local"]=parse(Float64,ARGS[3])
params["beta"]=(params["R_local"]*params["mu"])/(coloc_baseline[index_seeding,index_seeding]*N[index_seeding])
const T=size(coloc_3d,1)*7
const dt=1
const Tiles = size(coloc_3d,2)
const T_size = convert(Int32, round(T/dt))

@threads for i = 1:Iterations_multi
    R = zeros(Int32, T_size, Tiles)
    E = zeros(Int32, T_size, Tiles)
    I = zeros(Int32, T_size, Tiles)
    I[time_seeding,index_seeding]=1
    SEIRS_multi_patch!(E, I, R, T, dt, N, params, coloc_3d)
    integer_part=convert(Int32,floor(params["R_local"]))
    decimal_part=10*round(params["R_local"]-floor(params["R_local"]),digits=2)
    decimal_part_10=convert(Int32,floor(decimal_part))
    decimal_part_100=convert(Int32,10*round((decimal_part-floor(decimal_part)),digits=1))
    file_path = "$output_address/I+E_output_$(i)_R_local_$(integer_part)_$(decimal_part_10)"*"$(decimal_part_100).txt"
    file_path_1 = "$output_address/R_output_$(i)_R_local_$(integer_part)_$(decimal_part_10)"*"$(decimal_part_100).txt"
    writedlm(file_path, I.+E , ',')
    writedlm(file_path_1, R , ',')
    println("Writing part $i successful with R local $(params["R_local"])")
end

ending=time()
println(ending-starting)

File: Parallelized_SEIR_model.cpp
#include <iostream>
#include <random>
#include <vector>
#include <fstream>
#include <string>
#include <sstream>
#include <iomanip>
//#include <omp.h> // Include the OpenMP header


typedef std::vector<std::vector<double>> colocation_type;  // NxN matrix
typedef std::vector<int> count_vector;  // N vector of integers
typedef std::vector<count_vector>  run_container; //  T vectors of N 



double compute_p_beta(int i, double beta, double dt, int N, const colocation_type& coloc, const count_vector& I) {
    double p = 0.;
    for (int j=0; j<N; j++) {
        p += coloc[i][j] * double(I[j]);
    }
    return 1. - exp(-beta*dt*p);
}




int main(int argc, char* argv[]) {
    if (argc != 7) {
        std::cerr << "Wrong command-line arguments provided." << std::endl;
        return 1;
    }

    // read the integer parameters
    std::vector<int> parameters_int(6);
    std::ifstream read_param_int(argv[5]);
    if (!read_param_int.is_open()) {
            std::cerr << "Error opening file: " << argv[5] << std::endl;
        }
    for (int i = 0; i < 6; ++i) {
        if (!(read_param_int >> parameters_int[i])) {
                    std::cerr << "Error reading from file: " << argv[5] << std::endl;
                }
    }  // end i
    read_param_int.close();

    // read the float parameters
    std::vector<double> parameters_double(2);
    std::ifstream read_param_double(argv[4]);
    if (!read_param_double.is_open()) {
            std::cerr << "Error opening file: " << argv[4] << std::endl;
        }
    for (int i = 0; i < 2; ++i) {
        if (!(read_param_double >> parameters_double[i])) {
                    std::cerr << "Error reading from file: " << argv[4] << std::endl;
                }
    }  // end i
    read_param_double.close();

    // read the address list
    std::vector<std::string> address_file(2);
    std::ifstream read_param_address(argv[3]);
    if (!read_param_address.is_open()) {
            std::cerr << "Error opening file: " << argv[3] << std::endl;
        }
    for (int i = 0; i < 2; ++i) {
        if (!(read_param_address >> address_file[i])) {
                    std::cerr << "Error reading from file: " << argv[3] << std::endl;
                }
    }  // end i
    read_param_address.close();
    
    // PARAMETERS
    double R_local = std::stod(argv[1]);
    double alpha = parameters_double[0];
    double mu = parameters_double[1];
    int N = parameters_int[0];  // number of patches
    int n_runs = std::stod(argv[2]);  // number of stochastic runs
    int T = parameters_int[1];  // max time (days)
    int T_coloc = parameters_int[2];  // max colocation
    int t_start = parameters_int[3];  // day to start the simulation
    int index_seed = parameters_int[4];  // index of the patch in which the infection starts
    int initial_infected = parameters_int[5];  // number of initial infected
    std::string file_coloc_template = address_file[0];  // directory of the colocation files up to the index. eg. "./baseline/" 
    std::string file_pop = address_file[1]; // filename and path of the pop file
    std::string output_address = argv[6]; //

    // FIXED PARAMETERS
    int colocation_periodicity = 7;   // new colocation every xx days
    double dt = 1.0;  // discretization time (=1 if time is in days)


    ////////
    // number of available threads
    //int n_cpu = omp_get_num_threads();
    //std::cout << "Number of threads: " << n_cpu << std::endl;

    ////////
    // read the colocation
    std::vector<colocation_type> colocation(T_coloc, colocation_type(N, std::vector<double>(N, 0.0)));   // coloc[t][i][j] where t=colocation time index and i,j patches 
    colocation_type colocation_baseline(N, std::vector<double>(N, 0.0));

    std::string file_coloc = file_coloc_template + "coloc_matrix_baseline.txt";
        std::ifstream file(file_coloc);
        if (!file.is_open()) {
            std::cerr << "Error opening file: " << file_coloc << std::endl;
            return 1;
	}

        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                if (!(file >> colocation_baseline[i][j])) {
                    std::cerr << "Error reading from file: " << file_coloc << std::endl;
                    return 1;
		}
            }
        }
    for (int c=0; c<T_coloc; c++) {
        std::string file_coloc = file_coloc_template + "coloc_matrix_" + std::to_string(c) + ".txt";  // eg. "./baseline/coloc_3.txt"
        std::ifstream file(file_coloc);
        if (!file.is_open()) {
            std::cerr << "Error opening file: " << file_coloc << std::endl;
            return 1;
	}

        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                if (!(file >> colocation[c][i][j])) {
                    std::cerr << "Error reading from file: " << file_coloc << " at position (" << i << ", " << j << ")" << std::endl;
                    return 1;
		}
            }  // end for j
        } // end for i

        file.close();

    }  // end for c

    ////////
    // read the population
    std::vector<int> population(N);
    std::ifstream rin(file_pop);
    if (!rin.is_open()) {
            std::cerr << "Error opening file: " << file_pop << std::endl;
        }
    for (int i = 0; i < N; ++i) {
        if (!(rin >> population[i])) {
                    std::cerr << "Error reading from file: " << file_pop << std::endl;
                }
    }  // end i
    rin.close();

    double beta = (R_local*mu)/(colocation[0][index_seed][index_seed]*population[index_seed]);

    ////////////
    // base random seed
    unsigned int seed0 = static_cast<unsigned int>(time(0));  


    ////////
    // compute the probas that I can compute
    double p_mu = 1.-exp(-mu*dt);
    double p_alpha = 1.-exp(-alpha*dt);

    std::vector<run_container> attack_rate(n_runs, run_container(T, count_vector(N, 0)) );


    #pragma omp parallel for schedule(dynamic) // Parallelize the loop
    for (int it=0; it<n_runs; it++) {

        std::mt19937 gen(seed0+it);  // initialize the random seed, making sure no run has the same seed

        std::vector<count_vector> lE(T, count_vector(N, 0));  // lE[t][i]
        std::vector<count_vector> lI(T, count_vector(N, 0));
        std::vector<count_vector> lR(T, count_vector(N, 0));
        int E_new, I_new, R_new, S;
        double p_beta;
        int n_active_people;

        // initialize outbreak
        lI[t_start][index_seed] = initial_infected;
        attack_rate[it][t_start][index_seed] = initial_infected;

        // loop over times
        for (int t=t_start+1; t<T; t++) {

            // loop over patches
            n_active_people = 0;
            for (int i=0; i<N; i++) {

                // E to I
                if (lE[t-1][i]>0) {
                    std::binomial_distribution<> dEI(lE[t-1][i], p_alpha);
                    I_new = dEI(gen); 
                } else {
                    I_new = 0;
                }

                // I to R
                if (lI[t-1][i]>0) {
                    std::binomial_distribution<> dIR(lI[t-1][i], p_mu);
                    R_new = dIR(gen); 
                } else {
                    R_new = 0;
                }

                // S to E
                S = population[i]-lE[t-1][i]-lI[t-1][i]-lR[t-1][i];
                if (S>0) {
                    p_beta = compute_p_beta(i, beta, dt, N, colocation[t/colocation_periodicity], lI[t-1]);
                    std::binomial_distribution<> dSE(S, p_beta);
                    E_new = dSE(gen);
                } else {
                    E_new = 0;
                }
                
                // update
                lE[t][i] = lE[t-1][i] - I_new + E_new;
                lI[t][i] = lI[t-1][i] - R_new + I_new;
                lR[t][i] = lR[t-1][i] + R_new;
                attack_rate[it][t][i] = lE[t][i]+lI[t][i]+lR[t][i];
                n_active_people += lE[t][i] + lI[t][i];
    
            }  // end for i

            // 
            if (n_active_people == 0) {
                // the outbreak has died out. Propagate the number of recovered and stop
                for (int t2=t+1; t2<T; t2++) {
                    for (int i=0; i<N; i++) {
                        attack_rate[it][t2][i] = attack_rate[it][t][i];
                    }
                }
                break;
            }


        }  // end for t
        
        int integer_part = static_cast<int>(std::floor(R_local));
        double decimal_part = std::round(100*(R_local - std::floor(R_local)));

        std::ostringstream file_path_stream;
        file_path_stream << "gzip - > " << output_address << "/I+E+R_output_" << it+1 << "_R_local_" << integer_part << "_"
                        << decimal_part << ".txt.gz";
        std::string file_path = file_path_stream.str();

        FILE *pipe = popen(file_path.c_str(), "w");

        for (int i=0; i<T; i++) { 
            for (int j=0; j<N-1; j++) {
                fprintf(pipe, "%d ", attack_rate[it][i][j]);
            }
            fprintf(pipe, "%d\n", attack_rate[it][i][N-1]);
        }

        pclose(pipe);

    }  // end for it


    return 0;
}
